cpp/src/decisiontree/levelalgo/levelkernel_regressor.cuh

```diff
@@ -401,3 +450,597 @@ __global__ void get_best_split_regression_kernel(
     }
   }
 }
+
+//This is the best bin finder at block level for each column using one pass MSE
+template <typename T>
+DI GainIdxPair bin_info_gain_regression_mse(const T sum_parent,
+                                            const T *sum_left,
+                                            const unsigned int *count_right,
+                                            const int count, const int nbins) {
+  GainIdxPair tid_pair;
+  tid_pair.gain = 0.0;
+  tid_pair.idx = -1;
+  for (int tid = threadIdx.x; tid < nbins; tid += blockDim.x) {
+    unsigned int right = count_right[tid];
+    unsigned int left = count - right;
+    if ((right != 0) && (left != 0)) {
+      T mean_left = sum_left[tid];
+      T mean_right = sum_parent - mean_left;
+      T mean_parent = sum_parent / count;
+      float info_gain = (float)MSEGain<T>::exec(count, left, right, mean_parent,
+                                                mean_left, mean_right);
+      if (info_gain > tid_pair.gain) {
+        tid_pair.gain = info_gain;
+        tid_pair.idx = tid;
+      }
+    }
+  }
+  return tid_pair;
+}
+
+//This is the best bin finder at block level for each column using two pass MAE
+template <typename T>
+DI GainIdxPair bin_info_gain_regression_mae(const T mae_sum_parent,
+                                            const T *mae_sum_left,
+                                            const T *mae_sum_right,
+                                            const unsigned int *count_right,
+                                            const int count, const int nbins) {
+  GainIdxPair tid_pair;
+  tid_pair.gain = 0.0;
+  tid_pair.idx = -1;
+  for (int tid = threadIdx.x; tid < nbins; tid += blockDim.x) {
+    unsigned int right = count_right[tid];
+    unsigned int left = count - right;
+    if ((right != 0) && (left != 0)) {
+      T mae_left = mae_sum_left[tid] / (T)left;
+      T mae_right = mae_sum_right[tid] / (T)right;
+      T mae_parent = mae_sum_parent / (T)count;
+      float info_gain = (float)MAEGain<T>::exec(mae_parent, mae_left, mae_right,
+                                                left, right, count);
+      if (info_gain > tid_pair.gain) {
+        tid_pair.gain = info_gain;
+        tid_pair.idx = tid;
+      }
+    }
+  }
+  return tid_pair;
+}
+
+//One pass best split using MSE
```

Comment:
General comment about all the kernels below. There seems to be quite a bit of repeated code blocks. Maybe the common/shared logical blocks of code across these kernels can be refactored into individual device methods to improve code readability?

Reply:
the regressor is a bit tricky to merge code, as we have different ways to calculate MSE and MAE hence, we have more code here. MSE is one pass formula and MAE has a more general formula.

Reply:
I see. Thanks